R/quick-theme.R

```diff
@@ -0,0 +1,210 @@
+#'   be assigned to the `$primary` (BS4) or `$brand-primary` (BS3) Sass
+#'   variables.
+#' @param secondary A color string for the secondary color; if not `NULL`, this
+#'   will be assigned to the `$secondary` and `$default` Bootstrap 4 Sass
+#'   variables. (This argument is not currently supported for Bootstrap 3.)
+#'
+#' @section Implementation notes:
+#'
+#' `bs_theme_quick` supports both Bootstrap 3 and 4 (and 4+3), but the
+#' implementation differs greatly.
+#'
+#' For Bootstrap 4, the vast majority of default colors are directly or
+#' indirectly based on the `$black`, `$white`, and `$gray-100` through
+#' `$gray-900` variables; or on the theme colors (primary, secondary, danger,
+#' warning, info, etc.). `bs_theme_quick` sets `$white` to the `bg` color,
+#' `$black` to the `fg` color, and interpolates the grays between them. If
+#' provided, the `accent` argument is used to set the `$primary` variable, and
+#' the `secondary` argument is used to set `$secondary` and `$default`
+#' variables.
+#'
+#' For Bootstrap 3, a similar set of `$black`, `$white`, and `$gray-darker`
+#' through `$gray-lighter` variables exist, and these are populated using the
+#' same strategy as with Bootstrap 4. However, unlike Bootstrap 4, in Bootstrap
+#' 3 many of the default colors are hard-coded hex colors, that also happen to
+#' be shades of gray. `bs_theme_quick` overrides these hard-coded values with
+#' colors interpolated between `bg` and `fg`.
+#'
+#' @examples
+#'
+#' bs_theme_new("4+3")
+#' bs_theme_quick(bg = "#000060", fg = "skyblue",
+#'   accent = "orange", secondary = "silver")
+#'
+#' # You can apply further customizations here if desired, e.g.:
+#' bs_theme_add_variables("success" = "#1D7732")
+#'
+#' if (interactive()) {
+#'   bs_theme_preview()
+#' }
+#'
+#' @export
+bs_theme_quick <- function(bg = "#FFFFFF", fg = "#000000",
+  accent = NULL, secondary = NULL) {
+
+  theme <- bs_theme_get()
+  if (is.null(theme)) {
+    stop("No bootstraplib theme is active (did you forget to call bs_theme_new()?)")
+  }
+
+  bs4_tags <- c("bootstraplib_version_4", "bootstraplib_version_4+3")
+  bs3_tags <- c("bootstraplib_version_3")
+
+  results <- if (any(bs4_tags %in% theme$tags)) {
+    bs4_theme_quick(bg, fg, accent, secondary)
+  } else if (any(bs3_tags %in% theme$tags)) {
+    bs3_theme_quick(bg, fg, accent, secondary)
+  } else {
+    stop("bs_theme_quick doesn't recognize the active version of Bootstrap")
+  }
+  bs_theme_add(results)
+}
+
+bs4_theme_quick <- function(bg, fg, accent, secondary) {
+  white <- htmltools:::parseCssColors(bg)
+  black <- htmltools:::parseCssColors(fg)
+
+  grays <- colorRamp(c(white, black), alpha = TRUE)(0:10/10)
+
+  if (any(grays[,4] != 255)) {
+    warning(call. = FALSE,
+      "bs_theme_quick does not respect alpha in `white` and `black` arguments"
+    )
+  }
+
+  white_yiq <- color_yiq(grays[1,1], grays[1,2], grays[1,3])
+  black_yiq <- color_yiq(grays[11,1], grays[11,2], grays[11,3])
+
```

Comment:
What do you think about warning if the contrast ratio is less than 7 and pointing to this website (I was thinking of doing something like this for the `autoTheme` feature as well)? https://webaim.org/resources/contrastchecker/

Reply:
Great idea! I don't think I'll have time to implement it though--at least not today, I only have a few more minutes of coding time!

What if we let this PR go through without it, and if/when you add it to autoTheme, then we add it here as well?

Reply:
Sounds good, I'd still like some more time to review, but if no major blockers I'll review and merge today